Add renderer documentation (#4870)
GregTechCEu/GregTechCEu-1.19: docs/content/Modpacks/Other-Topics/Dynamic-Renders.md

---
title: Dynamic Renders
---

Dynamic renders allow you to attach custom block entity rendering logic to a machine. They run client-side every frame and can read live machine state to animate or display information in the world.

### How it works

> A `DynamicRender` is a typed renderer that is attached to a machine model. You create a class that extends `DynamicRender<T, S>`, where `T` is the machine type it reads from and `S` is the renderer class itself. The renderer is then registered in `ClientProxy`, and attached to a machine definition via `.model(...).andThen(b -> b.addDynamicRenderer(...))`.

Reviewer: Again, that example at the end won't work.

Reviewer: A method reference is better here IMO.

        // recipeLogic.isWorking() is safe here because RecipeLogic.status is @SyncToClient
        ItemStack displayStack = machine.recipeLogic.isWorking()
                ? new ItemStack(Items.GREEN_WOOL)
                : new ItemStack(Items.RED_WOOL);

        poseStack.pushPose();
        poseStack.translate(0.5, 2.5, 0.5); // center of block, 2 blocks up
        float totalTick = machine.getLevel().getGameTime() + partialTick;
        poseStack.mulPose(new Quaternionf().rotateY(totalTick * Mth.TWO_PI / 60));
        poseStack.scale(0.5f, 0.5f, 0.5f);

        Minecraft.getInstance().getItemRenderer().renderStatic(
                displayStack, ItemDisplayContext.FIXED,
                LightTexture.FULL_BRIGHT, OverlayTexture.NO_OVERLAY,
                poseStack, buffer, machine.getLevel(), 0);
        poseStack.popPose();
    }

    @Override
    public boolean shouldRenderOffScreen(WorkableElectricMultiblockMachine machine) {
        return true;
    }

    @Override
    public AABB getRenderBoundingBox(WorkableElectricMultiblockMachine machine) {
        BlockPos pos = machine.getBlockPos();
        return new AABB(pos.offset(-1, 0, -1), pos.offset(2, 4, 2));
    }
}
```

> !!! warning "Client-side data must be synced"

Reviewer: ```suggestion
!!! warning "Server-side data must be synced"
```

>     `render(...)` runs on the render thread, so any machine state you read must be marked `@SyncToClient`. Using server-only fields will silently read stale or default values. `RecipeLogic.status` and `RecipeLogic.isActive` are both `@SyncToClient`, which is why `recipeLogic.isWorking()` works here. If you add fields to a custom machine and need them in a renderer, annotate them accordingly. See the [sync annotations reference](../../Development/Data-Sync-System/Annotations.md) for details.

Reviewer: ```suggestion
    `render(...)` runs on the render thread, so any machine state you read must be sent to clients, for exakple with `@SyncToClient` . Using server-only fields will silently read stale or default values. `RecipeLogic.status` and `RecipeLogic.isActive` are both `@SyncToClient`, which is why `recipeLogic.isWorking()` works here. If you add fields to a custom machine and want to use them in a renderer, annotate them accordingly. See the [sync annotations reference](../../Development/Data-Sync-System/Annotations.md) for details.
```

> !!! warning "The Render is global"

Reviewer: ```suggestion
!!! warning "Renders are global"
```

    Only one instance of the render class exists. Make sure to not store any machine- or instance-specific data in the class. There is also no lifecycle management for BEs going off-screen or being destroyed, so by putting the data in the render class you would have a bunch of stale values and memory leaks.

### Registering the type

> Register your renderer type in `ClientProxy.initializeDynamicRenders()`:

Reviewer: Addons can't exactly do this, you know?
this should have the common init event (or whatever the caller of ths function is, anyway) instead.

Please add a proper example of registering a custom dynamic render type, a code block with the appropriate event listener is OK.

```java
DynamicRenderManager.register(GTCEu.id("item_above_controller"), ItemAboveControllerRender.TYPE);
```

### Attaching to a machine

> Use `.model(...).andThen(...)` when defining the machine. If the machine previously used the `.workableCasingModel(...)` shorthand, expand it into `.model(createWorkableCasingMachineModel(...))` so you can chain `.andThen`:

Reviewer: Too many parentheses in the first code block, that won't work because `andThen` doesn't exist on MachineBuilder.

```java
public static final MultiblockMachineDefinition MY_MACHINE = REGISTRATE
    .multiblock("my_machine", WorkableElectricMultiblockMachine::new)
    // ... recipe type, pattern, etc. ...
    .modelProperty(GTMachineModelProperties.RECIPE_LOGIC_STATUS, RecipeLogic.Status.IDLE)
    .model(createWorkableCasingMachineModel(
            GTCEu.id("block/casings/solid/machine_casing_inert_ptfe"),
            GTCEu.id("block/multiblock/large_chemical_reactor"))
            .andThen(b -> b.addDynamicRenderer(DynamicRenderHelper::createItemAboveControllerRender)))
    .hasBER(true)
    .register();
```

!!! warning "hasBER is required"

>     You must call `.hasBER(true)` on any machine that uses a dynamic renderer. Without it, the block entity renderer is not created and nothing will be drawn.

Reviewer: Eh, not really, some covers use BERs too so all machines have them anyway. Good practice to have it specified, though, so keep this.

> ### Optional overrides

Reviewer: Mind describing the optional methods a bit... More? A 3-10 word sentence doesn't quite capture what they're for or how they work.

---
title: Dynamic Renders
---

Dynamic renders allow you to attach custom block entity rendering logic to a machine. They run client-side every frame and can read live machine state to animate or display information in the world.

### How it works

> A `DynamicRender` is a typed renderer that is attached to a machine model. You create a class that extends `DynamicRender<T, S>`, where `T` is the machine type it reads from and `S` is the renderer class itself. The renderer is then registered in `ClientProxy`, and attached to a machine definition via `.model([base model].andThen(b -> b.addDynamicRenderer(...)))`.

Reviewer 2: ```suggestion
A `DynamicRender` is a typed renderer that is attached to a machine model. You create a class that extends `DynamicRender<T, S>`, where `T` is the machine type it reads from and `S` is the renderer class itself. The renderer is registered during mod initialisation, and attached to a machine definition via `.model([base model].andThen(b -> b.addDynamicRenderer(...)))`.
```

    .model(createWorkableCasingMachineModel(
            GTCEu.id("block/casings/solid/machine_casing_inert_ptfe"),
            GTCEu.id("block/multiblock/large_chemical_reactor"))

>             .andThen(b -> b.addDynamicRenderer(new ItemAboveControllerRender())))

Reviewer: won't work, `addDynamicRender` takes in a `Supplier<DynamicRender<?, ?>>`
